### DISCUSSION TITLE
Possibility to duplicate a Role & Permissions

### DISCUSSION THREAD

**Discussion by botzill** [directus/directus#3008] [Feature Requests]
Hi.

I was wondering if you have plans to add an option that will allow us to duplicate Role & Permissions entry. Sometimes we want almost the same options but with some custom changes, some hidden fields, or other minor changes. Also, this copy can include as well copy of the collection and module navigation. 

Thx.

**Comment by benhaynes** [comment]
I'd like to see this added, seems useful given the complex setup for roles. I'll defer to @rijkvanzanten on how complex this is to implement...

**Reply by Michael-Nussbaumer** [reply]
I had a similar idea and found this discussion. Maybe my idea could be an addition to this one. For my idea there are maybe two different approaches.

1. Make it possible that users can have more than one role. Then merge the permissions  of both roles together.
2. Let the roles have a parent role and inherit the permissions from the parent role. (Example: You have 2 roles, `editor` and `publisher`. You can configure the access for the `editor` and select it as parent of the `publisher` role. Then `publisher` could maybe have just an additional role for `updating` the `status` field.

**Reply by rijkvanzanten** [reply]
> Then merge the permissions of both roles together.

This sounds easier than it is.. You run into thought problems like "Do you merge with AND or OR" from the parent on permissions, are fields visible to the parent visible to the user, is there a way to _exclude_ a parent item thing (eg "parent can read all, but only interns are locked out of _ _ _") etc

**Reply by Michael-Nussbaumer** [reply]
My idea would just be a simple merge of all access permissions you get from your roles. You would add a role as child if you need a role with more permissions. 

If you want to remove permissions, you just have to switch `parent` and `child`. In your case you have a role which have all rights except the one you want to remove. The child role would add the permissions.

**Reply by benhaynes** [reply]
This seems simple from one point of view, but would be a huge (breaking change) update that needs to be fully thought-through (we tried it in v8, and gave up due to complexity). It also could add a lot of confusion to the end-user set up experience... and I don't know what we gain (technically speaking). A few things to consider:

Roles have more than just CRUD on them, they include validation, defaults, navigation overrides, etc... none of which can be "merged". Also, if we're merging role permissions, we'd also want a place to see the "computed" values... so that's another interface that might _busy up_ the Roles page.

Not saying this isn't something we'd do... but it needs a lot of thinking still.

**Reply by uebmaster** [reply]
Forget the merge, i just want the same that the initial proposal say, the posibility of duplicate a role & permissions lol xD

**Reply by uebmaster** [reply]
this is the same that #2781 right?

**Reply by mfobert** [reply]
Even the ability to re-use custom permissions created on other roles for the same object would be beneficial.

**Comment by Pachat** [comment]
As i faced the same problem, I manage to do it this way, directly on the DB:

1. Retrieve (after creation) the `id`s of the new role and of the source role.
2. Run the following on the database:
```
DELETE FROM directus_permissions WHERE role="0cfce6ee-019f-4cda-b61f-a955f0c678e6"
INSERT INTO directus_permissions (
    "action",
    collection,
    fields,
    permissions,
    presets,
    role,
    validation)
SELECT
    "action",
    collection,
    fields,
    permissions,
    presets,
    "0cfce6ee-019f-4cda-b61f-a955f0c678e6",
    validation
FROM directus_permissions WHERE role="fd5b2cf5-1bd9-469f-8dc4-7f4d7a5a92f8"
```
where `fd5b2cf5-1bd9-469f-8dc4-7f4d7a5a92f8` is the source role's id
and `0cfce6ee-019f-4cda-b61f-a955f0c678e6` the target role's id

To have it directly in Directus would be great.

**Comment by NilsBaumgartner1994** [comment]
Would like to see that too

**Comment by w0kyj** [comment]
Appears to be duplicate of https://github.com/directus/directus/discussions/2781
